File: src/main/java/br/com/model/entities/classes/Endereco/Endereco.java
package br.com.model.entities.classes.Endereco;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;

import br.com.model.entities.interfaces.EnderecoInterface;
import lombok.Data;

@Entity
@Inheritance(strategy = InheritanceType.JOINED)
@Data
public class Endereco implements EnderecoInterface {
    @Id
    @GeneratedValue
    private Integer id;
    private String logradouro;
    private String cidade;
    private String bairro;
    private String complemento;
    private String cep;
    private String numero;
    private float coordenadaX;
    private float coordenaxaY;

    public Endereco(String logradouro, String cidade, String bairro, String complemento, String cep, String numero,
            float coordenadaX, float coordenaxaY) {
        setLogradouro(logradouro);
        setCidade(cidade);
        setBairro(bairro);
        setComplemento(complemento);
        setCep(cep);
        setNumero(numero);
        setCoordenadaX(coordenadaX);
        setCoordenaxaY(coordenaxaY);
    }

    public Endereco(Endereco endereco) {
        setId(endereco.getId());
        setLogradouro(endereco.getLogradouro());
        setCidade(endereco.getCidade());
        setBairro(endereco.getBairro());
        setComplemento(endereco.getComplemento());
        setCep(endereco.getCep());
        setNumero(endereco.getNumero());
        setCoordenadaX(endereco.getCoordenadaX());
        setCoordenaxaY(endereco.getCoordenaxaY());
    }

    public Endereco() {
    }

    @Override
    public String toString() {
        return "\n-- Endereço--\nID: " + getId() + "\nBairro: " + getBairro() + "\nCep: " + getCep() + "\nCidade: "
                + getCidade() + "\nComplemento: " + getComplemento() + "\nCoordenada-X:" + getCoordenadaX()
                + "\nCoordenaxa-Y:" + getCoordenaxaY() + "\nLogradouro: " + getLogradouro() + "\nNumero: "
                + getNumero();
    }
}

File: src/main/java/br/com/model/Endereco.kt
package br.com.model

import br.com.model.usuario.Usuario
import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.*

@Entity
@Inheritance(strategy = InheritanceType.JOINED)
data class Endereco(
    @field:ManyToOne @JoinColumn(name = "id_usuario") @field:JsonIgnore var usuario: Usuario,
    var logradouro: String,
    var cidade: String,
    var bairro: String,
    var complemento: String,
    var cep: String,
    var numero: String? = null,
    var coordenadaX: Float? = 0f,
    private var coordenadaY: Float? = 0f
) {
    @Id
    @GeneratedValue
    var id: Int? = null

    init {
        usuario.enderecos.add(this)
    }

    override fun toString(): String {
        return "-- Endereço--\nID: $id\nBairro: $bairro\nCep: $cep\nCidade: $cidade\nComplemento: $complemento\nCoordenada-X:$coordenadaX\nCoordenaxa-Y:$coordenadaY\nLogradouro: $logradouro\nNumero: $numero\nusuario: ${usuario.nome}"
    }
}
